src/runtime/metal.cpp

```diff
@@ -369,7 +369,10 @@ WEAK int halide_metal_acquire_context(void *user_context, mtl_device **device_re
 
     // If the device has already been initialized,
     // ensure the queue has as well.
-    halide_abort_if_false(user_context, (device == nullptr) || (queue != nullptr));
+    if (device != nullptr && queue == nullptr) {
+        error(user_context) << "halide_metal_acquire_context: device initialized but queue is not.\n";
+        return -1;
```

Comment:
Nit: do we have a set policy on returning -1 vs returning an error code?

Reply:
Given the prevalence (here and elsewhere) of just returning -1 or similar, apparently not :-/

(This is something we really should improve; at least making a typedef for the return codes and attempting to return error values. But that's probably a separate issue.)

Reply:
Currently, runtime functions that are only ever called by other runtime functions often return anything non-zero back upwards, and then the calling function selects an appropriate error code (e.g. if this acquire context failed while trying to do a device copy, we ultimately return a device_copy_failed error). Sometimes you need to know the calling context to return the appropriate error.

This convention sort of makes sense, but it's not very clear in the code that there even is a convention happening here. There's no reason this couldn't return an failed_to_acquire_context error or something, and then the calling function could replace the error code if it wants to.

Reply:
> runtime functions that are only ever called by other runtime functions

there's also not a clean demarcation here -- I will bet you a dollar that we can find 'public' runtime functions that shortcut out in this way too.

Reply:
The demarcation is supposed to be that the exposed functions in device_interface.cpp are the entrypoints, and return real error codes. The specific backend implementations, and the default implementations in device_interface.cpp don't need to.